crates/sui-core/src/authority/authority_per_epoch_store.rs

```diff
@@ -3151,12 +3025,36 @@ impl AuthorityPerEpochStore {
                     shared_object_congestion_tracker,
                 );
 
-                if let Some((deferral_key, _)) = deferral_info {
-                    debug!(
-                        "Deferring consensus certificate for transaction {:?} until {deferral_key:?}",
-                        certificate.digest(),
-                    );
-                    return Ok(ConsensusCertificateResult::Deferred(deferral_key));
+                if let Some((deferral_key, deferral_reason)) = deferral_info {
+                    if transaction_deferral_within_limit(
+                        &deferral_key,
+                        self.protocol_config()
+                            .max_deferral_rounds_for_congestion_control(),
+                    ) {
+                        debug!(
+                            "Deferring consensus certificate for transaction {:?} until {:?}",
+                            certificate.digest(),
+                            deferral_key
+                        );
+                        return Ok(ConsensusCertificateResult::Deferred(deferral_key));
+                    } else if let DeferralReason::SharedObjectCongestion(congested_objects) =
```

Comment:
can you reorder these ifs, i find this block very confusing. consider something like

```
let DeferralReason::SharedObjectCongestion(congested_objects) = deferral_reason else { panic!() };
if transaction_deferral_within_limit(
   ...
```

Reply:
Re structured the code with the newly added metrics tracking. Hope it's clearer now.